File: src/store.ts
import fs from 'node:fs/promises';
import path from 'node:path';
import { DATA_DIR } from './config.js';
import type { StoreShape } from './types.js';

const STORE_PATH = path.join(DATA_DIR, 'store.json');

const emptyStore = (): StoreShape => ({
  users: [],
  history: [],
  photoUsageRecords: [],
  openaiRequests: []
});

export async function ensureDataDirs() {
  await fs.mkdir(DATA_DIR, { recursive: true });
}

export async function readStore(): Promise<StoreShape> {
  await ensureDataDirs();
  try {
    const raw = await fs.readFile(STORE_PATH, 'utf8');
    const parsed = JSON.parse(raw) as StoreShape;
    return {
      users: Array.isArray(parsed.users) ? parsed.users : [],
      history: Array.isArray(parsed.history) ? parsed.history : [],
      photoUsageRecords: Array.isArray(parsed.photoUsageRecords) ? parsed.photoUsageRecords : [],
      openaiRequests: Array.isArray(parsed.openaiRequests) ? parsed.openaiRequests : []
    };
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code === 'ENOENT') {
      return emptyStore();
    }
    throw error;
  }
}

export async function writeStore(store: StoreShape) {
  await ensureDataDirs();
  const tempPath = `${STORE_PATH}.tmp`;
  await fs.writeFile(tempPath, JSON.stringify(store, null, 2), 'utf8');
  await fs.rename(tempPath, STORE_PATH);
}

export async function updateStore(mutator: (store: StoreShape) => void | Promise<void>) {
  const store = await readStore();
  await mutator(store);
  await writeStore(store);
  return store;
}

File: store.py
from __future__ import annotations

import json
import threading
from pathlib import Path
from typing import Any, Callable

from config import DATA_DIR

STORE_PATH = DATA_DIR / "store.json"
_lock = threading.Lock()


def empty_store() -> dict[str, Any]:
    return {
        "users": [],
        "history": [],
        "photoUsageRecords": [],
        "clientLogs": [],
        "openaiRequests": [],
    }


def ensure_data_dirs() -> None:
    DATA_DIR.mkdir(parents=True, exist_ok=True)


def read_store() -> dict[str, Any]:
    ensure_data_dirs()
    if not STORE_PATH.exists():
        return empty_store()
    with STORE_PATH.open("r", encoding="utf-8-sig") as file:
        parsed = json.load(file)

    store = empty_store()
    for key in store:
        value = parsed.get(key)
        store[key] = value if isinstance(value, list) else []
    return store


def write_store(store: dict[str, Any]) -> None:
    ensure_data_dirs()
    temp_path = Path(f"{STORE_PATH}.tmp")
    with temp_path.open("w", encoding="utf-8") as file:
        json.dump(store, file, ensure_ascii=False, indent=2)
    temp_path.replace(STORE_PATH)


def update_store(mutator: Callable[[dict[str, Any]], None]) -> dict[str, Any]:
    with _lock:
        store = read_store()
        mutator(store)
        write_store(store)
        return store
